File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = list()
        path_stack = list()
        node = root
        
        while node or len(path_stack) > 0:
            while node:
                res.append(node.val)
                path_stack.append(node)
                node = node.left
                
            if len(path_stack):
                node = path_stack.pop()
                node = node.right
        return res
        

File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type ListNode struct {
	Val  *TreeNode
	Next *ListNode
}
type Stack struct {
	Head *ListNode
}

func (s *Stack) Pop() *TreeNode {
	cur := s.Head
	s.Head = s.Head.Next
	cur.Next = nil
	return cur.Val
}
func (s *Stack) Push(val *TreeNode) {
	cur := &ListNode{
		Val:  val,
		Next: s.Head,
	}

	s.Head = cur
}
func (s *Stack) Empty() bool {
	return s.Head == nil
}

func preorderRecursiveTraversal(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}

	*res = append(*res, root.Val)
	preorderRecursiveTraversal(root.Left, res)
	preorderRecursiveTraversal(root.Right, res)
}

func preorderIterativeTraversal(root *TreeNode) []int {
	res := []int{}
	stack := &Stack{}
	cur := root
	for cur != nil || !stack.Empty() {
		for cur != nil {
			res = append(res, cur.Val)
			stack.Push(cur)
			cur = cur.Left
		}
		cur = stack.Pop()
		cur = cur.Right
	}
	return res
}

func preorderTraversal(root *TreeNode) []int {
	return preorderIterativeTraversal(root)
}
